[JIT][NNC] Add handling of strides to dynamic shape support.

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71173
* #71138
* #70656
* __->__ #70464
* #71144

Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

Commit:
Add handling of strides to dynamic shape support.

[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."




[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1171,7 +1171,7 @@ class TensorExprFuser {
     for (Node* fusion_group : fusion_groups) {
       VLOG(1) << "GenerateGuard for fusion group: " << *fusion_group;
-      if (!GenerateGuard(fusion_group, /*add_composed_op=*/true)) {
+      if (!GenerateGuard(fusion_group, /*add_composed_op=*/false)) {
```
TODO: make this `/*add_composed_op*/` into a flag for fusion, cc @navahgar

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```


[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```


[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```


[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```


[ghstack-poisoned]

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1416,7 +1501,8 @@ std::vector<CodeGen::CallArg> TensorExprKernel::prepareRunArgs(
   std::vector<CodeGen::CallArg> runArgs;
-  runArgs.reserve(inputs.size() + bufOutputs_.size());
+  runArgs.reserve(
+      inputs.size() + input_stride_args_.size() + bufOutputs_.size());
```
If the strides here are runtime args, shouldn't that be already part of `inputs`?

eellison replied:
they're not arguments to the graph, they're taken from the input tensors themselves

eellison commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -26,6 +26,23 @@ using OptNameList = c10::optional<std::vector<std::string>>;
+    const IntArrayRef strides) {
+  int n_dim = static_cast<int>(sizes.size());
+
+  if (n_dim == 0 || strides[n_dim - 1] != 1) {
```
yea, this was just factoring out this function, I didn't make any changes, but good point I'll update

eellison commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:
```diff
@@ -69,25 +69,124 @@ std::map<int64_t, Value*> InsertSymbolicShapesCompute(
+StrideInput summarizeOutputStrides(const TensorType& tt) {
+  auto strides = *tt.strides().concrete_sizes();
+  auto sizes = *tt.sizes().concrete_sizes();
+  // We only try to maintain output striding for channels last tensors,
```
What do you mean ?  In other cases, we will just make it contiguous

eellison replied:
we could add support for permuting to other ordering of dims but other than that would require more substantial changes

navahgar replied:
We might have to support more generic striding because Hao mentioned that they come up in SR models.

Anyways, the point I had here was just to assert in other cases, so that we are doing something incorrectly. If we can't figure out the other cases here, then no problem.

eellison replied:
generic striding is already handle for inputs.. it's just that we don't preserve striding for outputs if they aren't contiguous or channels-last

eellison commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:
```diff
@@ -309,49 +470,70 @@ RegisterOperators reg_guard({
                       tensor.device() != device ||
                       tensor.dtype() != expected_scalar_types[i]) ||
                   tensor.requires_grad()) {
-                std::cout << " DEVICE /DTYPE/ REQUIRES GRAD MISMATCH\n";
```
I'm wary of runtime VLOG's, because this will get executed on every execution. Even if debug is set to false, we have to make a call to look up the environment flag which is slow

eellison commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.h:
```diff
@@ -33,5 +34,24 @@ TORCH_API void runTensorExprDynamicGroup(
+  S_ONE, // STRIDE_ONE: packed
+  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
+  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
+  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
If it's arbitrary it becomes `S_AS_ARG`

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```


[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```


[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

eellison commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:
```diff
@@ -69,25 +69,124 @@ std::map<int64_t, Value*> InsertSymbolicShapesCompute(
+    // Contiguous
+  } else if (
+      dim > 0 && strides[dim] == strides[dim - 1] * sizes[dim - 1] &&
+      (stride_inputs[dim - 1 + stride_inputs_offset] != StrideInput::S_CONT)) {
```
The previous StrideInput may not exist if `dim ==0`, and we would be incurring an extra read of a value that may not be used

eellison replied:
imo that wouldn't be clearer to follow than the logic that exists currently

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

eellison commented:
With contiguous inputs
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(1, SS(-6), requires_grad=0, device=cpu)], striding_inputs_desc=[["TENSOR_CONT"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=1]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(1, 10, strides=[10, 1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-6], striding_inputs_desc=[["TENSOR_CONT"]], striding_outputs_desc=[["TENSOR_CONT"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
with prim::TensorExprGroup_0 = graph(%x : Double(1, SS(-6), requires_grad=0, device=cpu),
      %SS_6 : int):
  %3 : Double(1, SS(-6), strides=[10, 1], requires_grad=0, device=cpu) = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
  %1 : Double(1, SS(-6), requires_grad=0, device=cpu) = aten::sigmoid(%3) # test/test_jit_fuser_te.py:2027:0
  return (%1)
```
Transposed arg and 1
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-8), SS(-8), requires_grad=0, device=cpu)], striding_inputs_desc=[["S_ONE", "S_TRAN_CONT"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(10, 10, strides=[1, 10], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-8], striding_inputs_desc=[["S_ONE", "S_TRAN_CONT"]], striding_outputs_desc=[["S_ONE", "S_TRAN_CONT"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
```
As arg input
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-9), requires_grad=0, device=cpu)], striding_inputs_desc=[["S_AS_ARG"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = aten::__getitem__(%17, %18)
      %9 : Double(10, strides=[1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-9], striding_inputs_desc=[["S_AS_ARG"]], striding_outputs_desc=[["S_AS_ARG"]]](%x, %19)
      -> (%9)
    block1():
      %23 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %24 : Tensor = aten::sigmoid(%23) # test/test_jit_fuser_te.py:2027:0
      %25 : (Tensor) = prim::TupleConstruct(%24)
      %22 : Tensor = prim::TupleUnpack(%25)
      -> (%22)
  return (%13)
```
Channels last 
```
graph(%x : Tensor,
      %y : Tensor,
      %z : Tensor):
  %12 : bool = prim::TensorExprDynamicGuard[types=[Double(SS(-11), SS(-12), SS(-13), SS(-14), requires_grad=0, device=cpu)], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]]](%x)
  %13 : Tensor = prim::If(%12)
    block0():
      %17 : int[] = aten::size(%x)
      %18 : int = prim::Constant[value=0]()
      %19 : int = prim::Constant[value=1]()
      %20 : int = prim::Constant[value=2]()
      %21 : int = prim::Constant[value=3]()
      %22 : int = aten::__getitem__(%17, %18)
      %23 : int = aten::__getitem__(%17, %19)
      %24 : int = aten::__getitem__(%17, %20)
      %25 : int = aten::__getitem__(%17, %21)
      %9 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-14, -13, -12, -11], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]], striding_outputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]]](%x, %25, %24, %23, %22)
      -> (%9)
    block1():
      %29 : Tensor = aten::tanh(%x) # test/test_jit_fuser_te.py:2027:0
      %30 : Tensor = aten::sigmoid(%29) # test/test_jit_fuser_te.py:2027:0
      %31 : (Tensor) = prim::TupleConstruct(%30)
      %28 : Tensor = prim::TupleUnpack(%31)
      -> (%28)
  return (%13)
```

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

navahgar reviewed (approved):
Mostly LGTM. Couple of minor comments below.

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).

Commit:
Update on "[JIT][NNC] Add handling of strides to dynamic shape support."


Add handling of strided input tensors to dynamic fusion. This is done with the same set of input striding specializations as https://github.com/pytorch/pytorch/pull/60684/: 
```
  S_ONE, // STRIDE_ONE: packed
  S_CONT, // STRIDE_CONTIGUOUS: stride[i + 1] * sizes[i + 1]
  S_TRAN_CONT, // STRIDE_TRANSPOSED_CONTIGUOUS: stride[i-1] * sizes[i-1]
  S_AS_ARG, // STRIDE_AS_ARG: stride passed in as runtime value
```
and then two additional specializations for a) contiguous tensor and b) channels-last tensor. channels-last is a common case and we should optimize for it. additionally, tensors natively store whether they are contiguous/channels-last contiguous, which makes it faster to check if tensors follow this pattern. 

Output striding will be done in a follow up. 

The striding is stored on both the TensorGroup node and on the guard node. The striding descriptors are stored as a vector of strings on the node for debugability and to make use of storing ivalues as attributes on nodes. 

As an example:


```

%8 : Double(10, 11, 12, 13, strides=[1716, 1, 143, 11], requires_grad=0, device=cpu) = prim::TensorExprGroup_0[symbolic_shape_inputs=[-37, -36, -35, -34], striding_inputs_desc=[["TENSOR_CONT_CHANNELS_LAST"]](%x, %24, %23, %22, %21)```
```

Differential Revision: [D33458649](https://our.internmc.facebook.com/intern/diff/D33458649)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458649).
